lib/create-contentful-api.ts

```diff
@@ -30,118 +30,75 @@ import { FieldsType } from './types/query/util'
+  getAssets<Locales extends LocaleCode = LocaleCode>(
+    query?: Omit<AssetQueries<AssetFields>, 'locale'>
+  ): Promise<AssetCollection<'WITH_ALL_LOCALES', Locales>>
+}
 
-    parseEntries<Fields extends FieldsType = FieldsType, Locales extends LocaleCode = LocaleCode>(
-      data: EntryCollection<Fields, 'WITHOUT_LINK_RESOLUTION' | 'WITH_ALL_LOCALES', Locales>
-    ): EntryCollection<Fields, 'WITH_ALL_LOCALES' | 'WITHOUT_LINK_RESOLUTION', Locales>
-  }
+type ClientMethodsWithoutAllLocales<Modifiers extends ChainModifiers> = {
+  withAllLocales: Client<AddChainModifier<Modifiers, 'WITH_ALL_LOCALES'>>
+  getEntry<Fields extends FieldsType>(
+    id: string,
+    query?: EntryQueries
+  ): Promise<Entry<Fields, Modifiers>>
 
-export type ClientWithLinkResolutionAndWithoutUnresolvableLinks = BaseClient &
-  BaseClientWithAssets & {
-    withAllLocales: ClientWithAllLocalesAndWithLinkResolutionAndWithUnresolvableLinks
-    getEntry<Fields extends FieldsType>(
-      id: string,
-      query?: EntryQueries
-    ): Promise<Entry<Fields, 'WITHOUT_UNRESOLVABLE_LINKS'>>
-    getEntries<Fields extends FieldsType>(
-      query?: EntriesQueries<Fields>
-    ): Promise<EntryCollection<Fields, 'WITHOUT_UNRESOLVABLE_LINKS'>>
-
-    parseEntries<Fields extends FieldsType = FieldsType>(
-      data: EntryCollection<Fields, 'WITHOUT_LINK_RESOLUTION'>
-    ): EntryCollection<Fields, 'WITHOUT_UNRESOLVABLE_LINKS'>
-  }
+  getEntries<Fields extends FieldsType>(
+    query?: EntriesQueries<Fields>
+  ): Promise<EntryCollection<Fields, Modifiers>>
 
-export type ClientWithAllLocalesAndWithLinkResolutionAndWithoutUnresolvableLinks = BaseClient &
-  BaseClientWithAssetsWithAllLocales & {
-    getEntry<Fields extends FieldsType, Locales extends LocaleCode = any>(
-      id: string,
-      query?: EntryQueries & { locale?: never }
-    ): Promise<Entry<Fields, 'WITH_ALL_LOCALES' | 'WITHOUT_UNRESOLVABLE_LINKS', Locales>>
+  parseEntries<Fields extends FieldsType = FieldsType>(
+    data: EntryCollection<Fields, 'WITHOUT_LINK_RESOLUTION'>
+  ): EntryCollection<Fields, Modifiers>
 
-    getEntries<Fields extends FieldsType, Locales extends LocaleCode = any>(
-      query?: EntriesQueries<Fields> & { locale?: never }
-    ): Promise<EntryCollection<Fields, 'WITH_ALL_LOCALES' | 'WITHOUT_UNRESOLVABLE_LINKS', Locales>>
+  getAsset(id: string, query?: { locale?: string }): Promise<Asset<undefined>>
 
-    parseEntries<Fields extends FieldsType = FieldsType, Locales extends LocaleCode = LocaleCode>(
-      data: EntryCollection<Fields, 'WITHOUT_LINK_RESOLUTION' | 'WITH_ALL_LOCALES', Locales>
-    ): EntryCollection<Fields, 'WITH_ALL_LOCALES' | 'WITHOUT_UNRESOLVABLE_LINKS', Locales>
-  }
+  getAssets(query?: AssetQueries<AssetFields>): Promise<AssetCollection<undefined>>
+}
 
-export type DefaultClient = ClientWithLinkResolutionAndWithUnresolvableLinks
+export type Client<Modifiers extends ChainModifiers> = BaseClient &
```

Comment:
Could you explain how this works? Especially why do we use `{}` when link resolution and unresolvable links modifiers are provided? Does it mean the returned type is just `BaseClient`?

Reply:
If the result is `{}` it means that we’re not adding methods in this section so the result will be a combination of `BaseClient` and the last part of the expression (with or without locales).
If you find a nicer way to do this I’m happy to clean this up. The only other way I know is setting the properties to `never` but that’s IMO even messier.

Reply:
Oh, I wish I knew a nicer way, IMO it's good as it is. Thanks for explaining!